src/routes/safe/components/Apps/communicator.ts

```diff
@@ -0,0 +1,103 @@
+import { MutableRefObject, useEffect, useState } from 'react'
+import {
+  getSDKVersion,
+  SDKMessageEvent,
+  MethodToResponse,
+  Methods,
+  ErrorResponse,
+  MessageFormatter,
+} from '@gnosis.pm/safe-apps-sdk'
+import { SafeApp } from './types.d'
+
+type MessageHandler = (
+  msg: SDKMessageEvent,
+) => void | MethodToResponse[Methods] | ErrorResponse | Promise<MethodToResponse[Methods] | ErrorResponse | void>
+
+class AppCommunicator {
+  private iframe: HTMLIFrameElement
+  private handlers = new Map<Methods, MessageHandler>()
+  private app: SafeApp
+
+  constructor(iframeRef: MutableRefObject<HTMLIFrameElement>, app: SafeApp) {
+    this.iframe = iframeRef.current
+    this.app = app
+
+    window.addEventListener('message', this.handleIncomingMessage)
+  }
+
+  on = (method: Methods, handler: MessageHandler): void => {
+    this.handlers.set(method, handler)
+  }
+
+  private isValidMessage = (msg: SDKMessageEvent): boolean => {
+    const sameOrigin = msg.origin === window.origin
+    const knownOrigin = this.app.url.includes(msg.origin)
+
+    return knownOrigin && !sameOrigin
+  }
+
+  private canHandleMessage = (msg: SDKMessageEvent): boolean => {
+    return Boolean(this.handlers.get(msg.data.method))
+  }
+
+  send = (data, requestId, error = false): void => {
+    const sdkVersion = getSDKVersion()
+    const msg = error
+      ? MessageFormatter.makeErrorResponse(requestId, data, sdkVersion)
+      : MessageFormatter.makeResponse(requestId, data, sdkVersion)
+
+    this.iframe.contentWindow?.postMessage(msg, this.app.url)
+  }
+
+  handleIncomingMessage = async (msg: SDKMessageEvent): Promise<void> => {
+    const validMessage = this.isValidMessage(msg)
+    const hasHandler = this.canHandleMessage(msg)
+
+    if (validMessage && hasHandler) {
+      const handler = this.handlers.get(msg.data.method)
```

Comment:
`this.handlers.get(msg.data.method) as MessageHandler` perpahs avoids using `@ts-expect-error`

Reply:
why should it be avoided?

Reply:
It may provide additional security if his handler gets removed accidentally